crates/maple_cli/src/cmd/tagfiles.rs

```diff
@@ -0,0 +1,269 @@
+use crate::cmd::cache::{cache_exists, send_response_from_cache, CacheEntry, SendResponse};
+use anyhow::Result;
+use filter::{matcher::LineSplitter, Source};
+use itertools::Itertools;
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Serialize};
+use std::fmt::Write;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::{Path, PathBuf};
+use structopt::StructOpt;
+use utility::clap_cache_dir;
+use anyhow::anyhow;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct TagInfo {
+    name: String,
+    path: String,
+    pattern: String,
+    kind: String,
+}
+
+impl TagInfo {
+    pub fn format(&self, cwd: &PathBuf, winwidth: usize) -> String {
+        static HOME: OnceCell<Option<PathBuf>> = OnceCell::new();
+
+        let name = format!("{} ", self.name);
+        let taken_width = name.len() + 1;
+        let path_len = self.path.len() + 2;
+        let mut adjustment = 0;
+
+        let mut home_path = PathBuf::new();
+        let path = Path::new(&self.path);
+        let path = path.strip_prefix(cwd).unwrap_or(
+            HOME.get_or_init(|| dirs::home_dir())
+                .as_deref()
+                .map(|home| {
+                    path.strip_prefix(home)
+                        .map(|path| {
+                            home_path.push("~");
+                            home_path = home_path.join(path);
+                            home_path.as_path()
+                        })
+                        .unwrap_or(path)
+                })
+                .unwrap_or(path),
+        );
+        let path = path.display();
+
+        let path_label = if taken_width > winwidth {
+            format!("[{}]", path)
+        } else {
+            let available_width = winwidth - taken_width;
+            if path_len > available_width && available_width > 3 {
+                let diff = path_len - available_width;
+                adjustment = 2;
+                let path = path.to_string();
+                let start = path.char_indices().nth(diff + 2).map(|x| x.0).unwrap_or(path.len());
+                let path = path[start..].to_string();
+                format!("[…{}]", &path)
+            } else {
+                format!("[{}]", path)
+            }
+        };
+
+        let path_len = path_label.len();
+        let text_width = if path_len < winwidth {
+            winwidth - path_len
+        } else {
+            winwidth
+        } + adjustment;
+
+        format!(
+            "{text:<text_width$}{path_label}:::{path}:::{pattern}",
+            text = name,
+            text_width = text_width,
+            path_label = path_label,
+            path = self.path,
+            pattern = self.pattern,
+        )
+    }
+
+    pub fn parse(base: &PathBuf, input: &str) -> Result<Self, std::string::ParseError> {
```

Comment:
Since we are going to implement a parser to parse tags line, we might also want to take a look at the vimscript version of ctags parser, https://github.com/liuchengxu/vista.vim/blob/bcb50336188074c9cfc1aad4b463f4c0e698a68c/autoload/vista/parser/ctags.vim#L94 . It's theoratically fragile so I previously only support the ctags with JSON output in vim-clap instead of spending time doing this job, glad you eventually work on this :(.

Reply:
Yes :)